compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1109,16 +1129,20 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
 impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
     fn encode_attrs(&mut self, def_id: LocalDefId) {
         let tcx = self.tcx;
-        let mut is_public: Option<bool> = None;
-
-        let mut attrs = tcx
+        let mut state = AnalyzeAttrState {
+            is_exported: LazyCell::new(|| tcx.effective_visibilities(()).is_exported(def_id)),
```

Comment:
Is the lazy computation really useful? `effective_visibilities` should already have been computed, and `is_exported` is a single hashmap access + some projections.

Reply:
I didn't measure the effect of this optimization, but maybe @lqd and @GuillaumeGomez did in https://github.com/rust-lang/rust/pull/98450 (if it ever worked).

The optimization may be reasonable because every line doc comment is a separate attribute and there may be dozens or even hundreds of them, and query means potentially taking a lock (in parallel compiler?), and `LazyCell` is simple enough to use.

Reply:
We did and it was quite visible. I'd still recommend to run a perf check with and without it.

Reply:
I mean: compute once and make `is_exported` a simple `bool`.

Reply:
Sorry, misread. In this case a simple `bool` should be fine indeed.

Reply:
@cjgillot 
>I mean: compute once and make `is_exported` a simple `bool`.

That introduces a query dependency for every def id (as opposed to only def ids with doc string), that's something I did measure in https://github.com/rust-lang/rust/pull/107136 (but with a different query) and it causes a pretty large regression due to growing dependency graph (see https://github.com/rust-lang/rust/pull/107136#issuecomment-1399165021 and below).